gpii/node_modules/lifecycleManager/src/LifecycleManagerSession.js

```diff
@@ -71,18 +72,90 @@ var gpii = fluid.registerNamespace("gpii");
         }
     });
 
+    gpii.lifecycleManager.session.setAutoSavePrefs = function (existingAutoSavePrefs, preferences, path) {
+        var prefsValue = fluid.get(preferences, path);
+        if (prefsValue) {
+            fluid.set(existingAutoSavePrefs, path, prefsValue);
+        }
+        return existingAutoSavePrefs;
+    };
+
+    /**
+     * Filter the preferences by metadata defined in the preferences set. Only the preferences that are allowed to be
+     * autosaved will be returned.
+     * @param {Object} preferences - The updated preferences.
+     * @param {Object} metadata - The metadata that defined in the prefs set.
+     * @return {Object} - The preferences to be autosaved.
+     */
+    gpii.lifecycleManager.session.filterPrefsByAutoSave = function (preferences, metadata) {
+        var autoSaveScope;
+        fluid.each(metadata, function (meta) {
+            if (meta.type === "autosave") {
+                autoSaveScope = autoSaveScope || {};
+                fluid.each(meta.scope, function (oneScope) {
+                    var oneScopeArray = fluid.makeArray(oneScope);
+                    fluid.set(autoSaveScope, oneScopeArray, true);
+                });
+            }
+        });
+
+        // Preferences will not be auto saved if the "autosave" metadata is not defined
+        if (!autoSaveScope) {
+            return undefined;
+        }
+
+        var prefsForAutoSave = {};
+        fluid.each(autoSaveScope, function (autoSaveLvl1Value, autoSaveLvl1Path) {
```

Comment:
Rather than iterating over the scope, doesn't it make more sense to iterate over the preferences document? Since that is morally what we are transforming. This would also allow you to simplify the save function since you would have the preference value in your hands during the loop.
Also I don't think your isObjectSimple test is safe since there is no reason a preference value might not be an object. I suggest the following:

fluid.each(preferences, function (prefValue, prefKey) {
    var autoSaveValue = autoSaveScope[prefKey];
    if (prefKey.startsWith("http://registry.gpii.net/applications")) {
        fluid.each(prefValue, function (commonValue, commonKey) {
            var autoSaveCommon = autoSaveValue ? autoSaveValue[commonKey] : null;
            if (autoSaveCommon === true) {
                fluid.set(prefsForAutoSave, [prefKey, commonKey], commonValue);
            }
        });
    else if (autoSaveValue === true) {
        fluid.set(prefsForAutoSave, [prefKey], prefValue);
    }
});

seems more reliable and easy to interpret